Allow for Backup labeling

ttpcodes opened the pull request:
https://github.com/longhorn/longhorn/issues/563

This PR extends the `longhorn-manager API` to allow users to be able to label `Backups` of `Snapshots`.

Calls to the `snapshotBackup` action on a `Volume` will now properly parse the `Labels` field of a `SnapshotInput` in order to validate them, making sure that every `Label` key conforms to a `Kubernetes Qualified Name` and that every `Label` value is non-empty. Additionally, the validation will ensure that none of the provided `Label` keys are reserved keywords (currently `KubernetesStatus`, `RecurringJobLabel`, and `ranchervm-base-image`). These will be passed to the `Engine API`, which will then make the appropriate calls with the `map` of `Labels` on the `longhorn-engine` to create the `Backup` with those `Labels`.

Additionally, the `snapshot` subcommand of `longhorn-manager`, which is used in `RecurringJobs`, has been modified to read the `KubernetesStatus` property of a `Volume` during a `Snapshot`/`Backup`, marshal that into `JSON`, and then set it as an additional `Label` on the `Snapshot` or `Backup`. That means that `Snapshots` and `Backups` created by the `RecurringJob` functionality will now automatically have `KubernetesStatus` and `RecurringJob` set as `Labels`.

As an aside, because the `BackupSnapshot` method of `VolumeManager` is still calling `SnapshotBackup` of `Engine API` in a `goroutine`, any errors reported on `Backup` initialization by that call will not be passed back to the user during the `Request`, meaning that even if validation passed on the `longhorn-manager` end and the user receives a `200 OK` response from the server, the `Backup` may not be initialized. This should be fixed in a separate PR.

yasker commented on api/snapshot.go:
```diff
@@ -159,6 +161,25 @@ func (s *Server) SnapshotBackup(w http.ResponseWriter, req *http.Request) (err e
+		labels[key] = val
+	}
+
+	reserved := types.ReservedSnapshotLabels()
```
Just use a map for reserved key check.

yasker commented on app/volume.go:
```diff
@@ -129,6 +130,13 @@ func NewJob(volumeName, snapshotName, backupTarget string, labels map[string]str
+	// Save current KubernetesStatus of Volume as a Label on the Snapshot/Backup.
+	kubeStatus, err := json.Marshal(v.Status.KubernetesStatus)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not convert volume %v's KubernetesStatus to json", volumeName)
```
`BUG:`

ttpcodes commented:
- Moved the reserved `Label` keys into a single `slice` instead of generating a new one every time on `method` call.
- Moved validation of `Labels` into the `util` package for other methods to use.
- Marked failed marshal of `KubernetesStatus` as a `BUG` in the log.
- Allowed user to specify `Labels` on a `RecurringJob`.
- Attempt to restore `KubernetesStatus` on `Backup` restore, with the `LastPodRefAt` and `LastPVCRefAt` fields set to `SnapshotCreated`.
- Modified `ParseLabels` method to use the same implementation as `longhorn-engine`.

ttpcodes commented:
- Check that `KubernetesStatus` isn't a `zero` value before setting it on the `Backup`.
- Set `KubernetesStatus` on only `Backups` and not `Snapshots` for `RecurringJobs`.

yasker commented on manager/volume.go:
```diff
@@ -166,6 +168,16 @@ func (m *VolumeManager) Create(name string, spec *types.VolumeSpec) (v *longhorn
+		// If KubernetesStatus is set on Backup, restore it.
+		if statusJSON, ok := backup.Labels[types.KubernetesStatusLabel]; ok {
+			if err := json.Unmarshal([]byte(statusJSON), kubeStatus); err != nil {
+				return nil, err
```
Don't error out. Print an error log then proceed. It's not critical enough to stop restoration.

ttpcodes commented:
- Logged error with `JSON` unmarshal of `KubernetesStatus` instead of returning an `error` and stopping the rebuild.
- Prevented setting of `LastPodRefAt` and `LastPVCRefAt` if the `KubernetesStatus` could not be unmarshaled.

yasker commented:
LGTM. Need integration test.

ttpcodes commented:
- `LastPodRefAt` is now only set if the `WorkloadsStatus` field indicates that it had been referenced by a `Pod` in the past.
- `LastPVCRefAt` is now only set if the `PVCName` field indicates that it had been referenced by a `PersistentVolumeClaim` in the past.

yasker commented on manager/volume.go:
```diff
@@ -166,6 +168,23 @@ func (m *VolumeManager) Create(name string, spec *types.VolumeSpec) (v *longhorn
+					backup.Name, err)
+			} else {
+				// We were able to unmarshal KubernetesStatus. Set the Ref fields.
+				if kubeStatus.PVCName != "" {
```
`&& kubeStatus.LastPVCRefAt == ""`

Ditto below.

ttpcodes commented:
- `LastPodRefAt` and `LastPVCRefAt` now must also be already empty for them to be set manually.
